contracts/contracts/GrantRoundManager.sol

```diff
@@ -32,35 +24,70 @@ contract GrantRoundManager {
   /// @notice Address of the ERC20 token in which donations are made
   IERC20 public immutable donationToken;
 
+  /// @notice Used during donations to temporarily store swap output amounts
+  mapping(IERC20 => uint256) internal swapOutputs;
+
+  /// @notice Used during donations to temporarily store donation ratio sums to ensure totals are always 100%
+  mapping(IERC20 => uint256) internal donationRatios;
+
   /// @notice WETH address
-  address public constant WETH = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+  IERC20 public constant WETH = IERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+
+  /// @notice Scale factor
+  uint256 internal constant WAD = 1e18;
+
+  /// --- Types ---
+  /// @notice Defines the total `amount` of the specified `token` that needs to be swapped to `donationToken`. If
+  /// `path == donationToken`, no swap is required and we just transfer the tokens
+  struct SwapSummary {
+    uint256 amountIn;
+    uint256 amountOutMin; // minimum amount to be returned after swap
+    bytes path;
+  }
 
+  /// @notice Donation inputs and Uniswap V3 swap inputs: https://docs.uniswap.org/protocol/guides/swaps/multihop-swaps
+  struct Donation {
+    uint96 grantId; // grant ID to which donation is being made
+    IERC20 token; // address of the token to donate
+    uint256 ratio; // ratio of `token` to donate, specified as numerator where WAD = 1e18 = 100%
```

Comment:
for swaps with different decimals we could floor the amount for the tokens with the least amount of significant figures

Reply:
Seems related to the above comment, but I'm not sure what you mean here and where you need to floor things that isn't already floored by integer division in Solidity?

Reply:
maybe I'm missing something, but it's not related to the calculation per say but when we transfer an amount, emitting one value but really another has been sent to the contract

Reply:
The `_donationAmount` that is transferred is the same value as we emit in the `GrantDonation` method. Sorry, I don't think I understand the issue you are pointing out 😬

Reply:
@corydickson I'm just running the example in my head and i don't see an issue either.
When donate is invoked for a given token -> sum of ratios should be 1e18.

So if we have a token with 16 decimals and contributions are made to 3 diff grants -> 5 , 5, 10
The ratio would be 25%, 25%, 50%.
The frontend would backfill the last 2 decimals with 0 and send it. 


So once we do the swap to 18 decimal token -> you'd still have the same ratio of 25%, 25%, 50%  right ?

The only scenario where this might mess things up is if we had a token > 18 decimals  (which is not something we need to be bothered about)